### PR TITLE
Undo Redo: Add events to notify when a batch of actions starts and ends

### PR DESCRIPTION
**Pull request by EchterAgo**
This will make it possible to handle changing the undo redo stack more
efficiently by batching writes and/or combining multiple value changes to
the same value.

Signed-off-by: Axel Gembe <axel@gembe.net>

### CODE REVIEW

File: src/Gemini/Modules/UndoRedo/Services/UndoRedoManager.cs [hunk 1]
```diff
@@ -59,26 +81,56 @@ public void UndoAll()
...
-                Push(_undoStack, thisAction);
-                if (thisAction == action)
-                    return;
+                OnEnd();
             }
         }
 
+        private void OnBegin()
+        {
+            if (BatchBegin != null)
```
Comment (tgjones) [inline src/Gemini/Modules/UndoRedo/Services/UndoRedoManager.cs]: Please could you change these to use the thread-safe style of raising an event? i.e.

``` csharp
var handler = BatchBegin;
if (handler != null)
    handler(this, EventArgs.Empty);
```

Comment (EchterAgo) [inline src/Gemini/Modules/UndoRedo/Services/UndoRedoManager.cs]: Of course

**Comment by EchterAgo** [comment]
Raise event in a thread-safe style

**Comment by tgjones** [comment]
Thank you!

**Pull request merged**